File: py/counting_flips.py
import random

def simulate_game():
    # Generate 100 coins with random heads (1) or tails (0)
    coins = [random.choice([0, 1]) for _ in range(100)]
    
    # Alice and Bob both start the same
    alice_index = 0
    bob_index = 0
    alice_heads = 0
    bob_heads = 0

    # Alice's and Bob's turns
    while alice_index < 100 or bob_index < 100:
        # Alice's turn
        if alice_index < 100:
            if coins[alice_index] == 1:
                alice_heads += 1
            alice_index += 1
        
        # Bob's turn
        if bob_index < 100:
            if coins[bob_index] == 1:
                bob_heads += 1
            bob_index += 2  # Bob goes to the next odd index
        
        # Bob goes back to evens if he finishes odds
        if bob_index >= 100:
            bob_index = 1

        if alice_heads == 2 and bob_heads == 2:
            return "Tie"
        if bob_heads == 2:
            return "Bob"
        if alice_heads == 2:
            return "Alice"

    return "Tie"

def run_simulations(num_simulations):
    results = {"Alice": 0, "Bob": 0, "Tie": 0}
    
    for _ in range(num_simulations):
        result = simulate_game()
        results[result] += 1
    
    # Calculate percentages
    percentages = {key: (value / num_simulations) * 100 for key, value in results.items()}
    
    return percentages

# Run the simulation 100,000 times
num_simulations = 1000000
percentages = run_simulations(num_simulations)

print(f"After {num_simulations} simulations:")
print(f"Alice: {percentages['Alice']:.2f}%")
print(f"Bob: {percentages['Bob']:.2f}%")
print(f"Tie: {percentages['Tie']:.2f}%")

File: src/counting_flips.rs
use rand::random;
use std::collections::HashMap;
use std::fmt;

// Define an enum for the possible outcomes
#[derive(PartialEq, Eq, Hash)]
pub enum Outcome {
    Alice,
    Bob,
    Tie,
}

impl fmt::Display for Outcome {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let outcome_str = match self {
            Outcome::Alice => "alice",
            Outcome::Bob => "bob",
            Outcome::Tie => "tie",
        };
        write!(f, "{}", outcome_str)
    }
}

fn simulate_game(flips: usize) -> Outcome {
    // Generate 100 random coins (1 for heads, 0 for tails)
    let coins = random::<u128>();

    let found = (0..(flips / 2 - 1))
        .map(|v| v * 2)
        .chain((0..(flips / 2 - 1)).map(|v| v * 2 + 1))
        .zip(0..flips)
        .map(|(a, b)| (coins & (1 << a) > 0, coins & (1 << b) > 0))
        .scan((0, 0), |count, coin| {
            count.0 += coin.0 as u8;
            count.1 += coin.1 as u8;
            Some(*count)
        })
        .find(|count| count.0 == 2 || count.1 == 2);

    match found {
        None | Some((2, 2)) => Outcome::Tie,
        Some((2, _)) => Outcome::Bob,
        Some((_, 2)) => Outcome::Alice,
        _ => unreachable!(),
    }
}

pub fn run_simulations(num_simulations: usize) -> HashMap<Outcome, usize> {
    (0..num_simulations).fold(HashMap::new(), |mut hm, _| {
        hm.entry(simulate_game(10))
            .and_modify(|count| *count += 1)
            .or_insert(1);
        hm
    })
}
